### PR TITLE
Add keybinder example

### PR DESCRIPTION
**Pull request by ewnd9**
Related to #13 

Found a segmentation fault when clicking the second time on key binding, so added failing test.

Let me know if I need to adjust code style or can help with troubleshooting.

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -4,7 +4,7 @@ test_steps: &test_steps
   - checkout
   - run: echo "deb http://deb.debian.org/debian testing main" >> /etc/apt/sources.list
   - run: apt-get update
-  - run: apt-get install -y xvfb libgirepository1.0-dev libgtk-3-dev # xvfb for running headless tests (headless X11 server).
+  - run: apt-get install -y xvfb libgirepository1.0-dev libgtk-3-dev libkeybinder-3.0-0 gir1.2-keybinder xdotool # xvfb for running headless tests (headless X11 server).
```
Comment (ewnd9) [inline .circleci/config.yml]: A bit of a churn with additional dependencies, but I think global hotkeys is essential for desktop apps

**Comment by Place1** [comment]
Hey thanks for this. I'm a bit strapped for time but i'll get to this.

File: .circleci/config.yml (hunk 1 above)
Comment (Place1) [reply .circleci/config.yml]: I agree keybindings are important for desktop applications and i'm happy to keep the example code in the application as a demo for people to run.

I don't think we should include this dependency in the test suite though as it adds to the local environment setup for people looking to contribute. `node-gir` runs on MacOS and should only depend on `gobject-introspection` and `nodejs`. One could argue we shouldn't have included GTK as a testing dependency but I'm not familiar enough with core glib/gobject-introspection to write all the tests against just what's available in those libraries; and so GTK was an obvious choice as it's available on all platforms (including MacOS via `brew install gtk+3`) and ships in most linux distros by default.

So i'd prefer not to add this dependency, but I definitely want to keep the eample app.

File: __tests__/examples.test.js [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+describe('examples should work', () => {
+  test('examples/keybinder.js', async done => {
+    const proc = execFile('node', [path.resolve(`${__dirname}/../examples/keybinder.js`)], (err, stdout, stderr) => {
+      expect(err.signal).toBe('SIGTERM');
+      expect(stdout).toBe('<Super>X true\nclicked\nclicked\n');
+      expect(stderr).toBe('');
+      done();
+    });
+
+    await new Promise(resolve => setTimeout(resolve, 1000));
```
Comment (Place1) [inline __tests__/examples.test.js]: These 1 second delays can cause problems in large test suites as the delays add up. The test suite runs in ~1.5sec locally for me and this test, by itself, is twice that due to the waiting. I don't think we should keep this in the test suite for reasons explained in my other comment https://github.com/Place1/node-gir/pull/14#discussion_r163795369

but I wanted to comment on these timeouts for future reference :)

File: .circleci/config.yml (hunk 1 above)
Comment (ewnd9) [reply .circleci/config.yml]: I see the following resolution: move the test file to a different directory (e.g. `__tests__/integration`), tweak jest config to ignore it by default, add a separate npm script `test:integration` which runs only that directory and a separate CircleCI build with additional dependencies.

What do you think?

Comment (Place1) [reply .circleci/config.yml]: not a bad idea. Would be good to create seperate setups for linux/mac/windows integration tests and document what people need to install to run them. not sure how many we'll write though as our existing unittests are basically integration tests (testing bindings from JS). we're more likely to start writing c++ tests to unittest core parts of the bindings codebase. I'm happy to accept the proposed change though if you want more than an addition to `examples/`

**Comment by Place1** [comment]
@ewnd9 if you allow maintainers to commit to this pull request I can make the changes

**Comment by ewnd9** [comment]
@Place1 Sorry for keeping you waiting. Updated now.

> Would be good to create seperate setups for linux/mac/windows integration tests

I suppose there would be different cases, `keybinder` library is `x11` based for example.

> document what people need to install to run them

Not sure what the wording should be for documentation (a full list of dependencies and tested distros or just a reference to the CI config or something else?), so left just a command in README.

**Comment by ewnd9** [comment]
FYI After forgotten rebase I see the following error in the main tests (introduced in https://github.com/Place1/node-gir/commit/768d369741abaca6c70c01fe180cd103f0b4d9fa)

```txt
 FAIL  __tests__/struct.test.js
  ● struct › constructors › can be constructed using properties object

    expect(received).toEqual(expected)
    
    Expected value to equal:
      10
    Received:
      0

      17 |       });
      18 |       expect(typeof(rectangle)).toEqual('object');
    > 19 |       expect(rectangle.width).toEqual(10);
      20 |       expect(rectangle.height).toEqual(20);
      21 |       expect(rectangle.x).toEqual(0);
      22 |       expect(rectangle.y).toEqual(-5);
      
      at Object.test (__tests__/struct.test.js:19:31)


 RUNS  __tests__/callbacks.test.js
*** Error in `node': free(): invalid pointer: 0x0000000002fccbb0 ***
======= Backtrace: =========
/lib/x86_64-linux-gnu/libc.so.6(+0x777e5)[0x7f53a6dd87e5]
/lib/x86_64-linux-gnu/libc.so.6(+0x8037a)[0x7f53a6de137a]
/lib/x86_64-linux-gnu/libc.so.6(cfree+0x4c)[0x7f53a6de553c]
/usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0(g_boxed_free+0x13b)[0x7f53a46c1e6b]
/home/ewnd9/github/gtk-node/place1-node-gir/build/Release/girepository.node(_ZN3gir9GIRStructD2Ev+0x99)[0x7f53a4b51c69]
/home/ewnd9/github/gtk-node/place1-node-gir/build/Release/girepository.node(_ZN3gir9GIRStructD0Ev+0x11)[0x7f53a4b51cb1]
node(_ZN2v88internal13GlobalHandles31DispatchPendingPhantomCallbacksEb+0xee)[0xe7b61e]
node(_ZN2v88internal13GlobalHandles31PostGarbageCollectionProcessingENS0_16GarbageCollectorENS_15GCCallbackFlagsE+0x2a)[0xe7b88a]
node(_ZN2v88internal4Heap24PerformGarbageCollectionENS0_16GarbageCollectorENS_15GCCallbackFlagsE+0x2be)[0xea6c0e]
node(_ZN2v88internal4Heap14CollectGarbageENS0_15AllocationSpaceENS0_23GarbageCollectionReasonENS_15GCCallbackFlagsE+0x3a1)[0xea79e1]
node(_ZN2v88internal7Factory13NewFixedArrayEiNS0_13PretenureFlagE+0x4e)[0xe4b4de]
```

**Comment by Place1** [comment]
Good point. Perhaps it’s not worth automating the examples. We’d have so
many for different environments. Perhaps we can focus on unit tests and
examples.

**Comment by ewnd9** [comment]
Automating isn't hard and one time task though, so I think it makes sense at least to start with basic examples for each platform for ensuring stability (right now there is only Linux/Debian).

With the separate builds it shouldn't bother too much.

**Comment by Place1** [comment]
ok let's do this :D

**Comment by Place1** [comment]
i've fixed this on master

**Comment by ewnd9** [comment]
Yay, working for me :tada: What do you want me to do with this PR now? Discard all the changes except `examples/keybinder.js`?

**Comment by Place1** [comment]
Yeah let’s keep the example for others to see 😀

**Comment by ewnd9** [comment]
@Place1 Updated

**Comment by Place1** [comment]
are we happy to merge this in @ewnd9 ?

**Comment by ewnd9** [comment]
Yep :+1:

**Pull request merged**